### PR TITLE
feat(eslint-plugin): add `no-useless-empty-export` rule

### LINKED ISSUE #2757: New rule: no-useless-empty-export
In TypeScript, modules are TypeScript files that contain an import or export statement. If a file is not a module, it will be ignored by TypeScript. Sometimes a file doesn’t contain imports or exports for legitimate reasons. In this case, an empty export must be added. I.e. to augment an interface:

```ts
export {};

declare module '@appsemble/sdk' {
  interface Parameters {
    interval: number;
  }

  interface EventEmitters {
    interval: never;
  }
}
```

As soon as the file is modified to contain actual imports or exports, the empty export can be removed.

I’d like to propose a rule that reports empty exports if the file contains other imports / exports. This can be autofixable by removing the empty export. I’m willing to implement this.

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

-   [x] Addresses an existing issue: fixes #2757
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Adds a fun little new rule.

Intentionally only targets the useless `export {}` case. Targeting accidental global pollution is a more nuanced issue that should probably get its own issue+discussion+rule.

### CODE REVIEW

File: packages/eslint-plugin/src/rules/no-useless-empty-export.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    messages: {
+      uselessExport: 'Empty export does nothing and can be removed.',
+    },
+    schema: [],
+    type: 'suggestion',
+  },
+  defaultOptions: [],
+  create(context) {
+    return {
+      Program(node): void {
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/no-useless-empty-export.ts:47]: nit - it'd be good if this handled `declare module`/`declare namespace`:
https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEBbA9sArhBAzZz4G8BYAKHjPhAA8AHZGAF3jGQDsBnRy+AXngEYA3CRLkKNOo3wBfIcSlA

note that without a `declare`, an empty `export {}` is semantically illegal!

Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/no-useless-empty-export.ts]: > note that without a declare, an empty export {} is semantically illegal!

Sorry, I don't follow - is that a change request?

Comment (bradzacher) [reply packages/eslint-plugin/src/rules/no-useless-empty-export.ts]: No that's just an interesting titbit about the language that I learned double checking the syntax!

File: packages/eslint-plugin/tests/rules/no-useless-empty-export.test.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    'export = 3;',
+    'export const _ = {};',
+    `
+      const _ = {};
+      export default _;
+    `,
+    `
+      export * from '_';
+      export = {};
+    `,
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/rules/no-useless-empty-export.test.ts:35]: nit for sanity checking
```suggestion
    `,
    `
      export {};
    `,
```

**Review by bradzacher — APPROVED** [review]
LGTM - one comment.
Feel free to merge once addressed!

**Comment by JoshuaKGoldberg** [comment]
@bradzacher, sorry, the review got wiped. Could you ✔️ again please? 😄  

Same with https://github.com/typescript-eslint/typescript-eslint/pull/4378

**Review by bradzacher — APPROVED** [review]
feel free to use your admin rights to merge if i've already stamped.
you needn't wait for a re-review for trivial changes :)

**Comment by JoshuaKGoldberg** [comment]
I... don't think I have those admin rights? Maybe it's an auto-merge weirdness. I'll ping you separately if I need. Thanks!

**Comment by JoshuaKGoldberg** [comment]
Ah, this: https://github.com/typescript-eslint/typescript-eslint/pull/4503#issuecomment-1049333694

**Pull request merged**